crates/runtime/src/embeddings/index/query_table.rs

```diff
@@ -555,3 +557,462 @@ fn metadata_columns_to_exprs(metadata_columns: &MetadataColumns) -> Vec<Expr> {
+        ) -> datafusion::error::Result<datafusion::execution::SendableRecordBatchStream> {
+            self.0.execute(partition, context)
+        }
+    }
+
+    impl DisplayAs for ExplainExecutionPlan {
+        fn fmt_as(
+            &self,
+            _t: datafusion::physical_plan::DisplayFormatType,
+            f: &mut std::fmt::Formatter,
+        ) -> std::fmt::Result {
+            let columns: Vec<String> = self
+                .schema()
+                .fields()
+                .iter()
+                .map(|f| f.name())
+                .cloned()
+                .collect();
+
+            write!(
+                f,
+                "ExplainExecutionPlan: projection={columns:?} filter={:?} limit={:?}",
+                self.1, self.2,
+            )?;
+            Ok(())
+        }
+    }
+
+    #[async_trait::async_trait]
+    impl TableProvider for ExplainMemTable {
+        fn as_any(&self) -> &dyn Any {
+            self
+        }
+
+        fn schema(&self) -> SchemaRef {
+            self.0.schema()
+        }
+
+        fn table_type(&self) -> TableType {
+            self.0.table_type()
+        }
+
+        async fn scan(
+            &self,
+            state: &dyn Session,
+            projection: Option<&Vec<usize>>,
+            filters: &[Expr],
+            limit: Option<usize>,
+        ) -> Result<Arc<dyn ExecutionPlan>, DataFusionError> {
+            Ok(Arc::new(ExplainExecutionPlan(
+                self.0.scan(state, projection, filters, limit).await?,
+                filters.to_vec(),
+                limit,
+                projection.cloned(),
+            )) as Arc<dyn ExecutionPlan>)
+        }
+
+        fn supports_filters_pushdown(
+            &self,
+            filters: &[&Expr],
+        ) -> Result<Vec<TableProviderFilterPushDown>, DataFusionError> {
+            Ok(vec![TableProviderFilterPushDown::Exact; filters.len()])
+        }
+    }
+
+    /// An implementation of [`VectorIndex`] that has one row. Useful for testing explain plans.
+    #[derive(Debug)]
+    pub struct PretendVectorIndex {
+        embedded_column: String,
+        primary_columns: Vec<Field>,
+        schema: Schema,
+    }
+
+    #[async_trait::async_trait]
+    impl VectorIndex for PretendVectorIndex {
+        fn embedded_column(&self) -> String {
+            self.embedded_column.clone()
+        }
+
+        fn primary_fields(&self) -> Vec<Field> {
+            self.primary_columns.clone()
+        }
+
+        fn list_table_provider(&self) -> Arc<dyn TableProvider> {
+            Arc::new(ExplainMemTable(
+                MemTable::try_new(
+                    Arc::new(self.schema.clone()),
+                    vec![vec![one_row_default_record_batch_for_schema(&Arc::new(
+                        self.schema.clone(),
+                    ))]],
+                )
+                .unwrap(),
+            ))
+        }
+
+        fn metadata_columns(&self) -> MetadataColumns {
```

Comment:
You could compute these metadata column on init and then you don't need to return an owned clone